src/common.jl

```diff
@@ -1,269 +1,250 @@
-# varying element type. Another lisp-y recursion trick fixes that
-has_colon(idx::Tuple{}) = false
-has_colon(idx::Tuple) = isa(first(idx), Colon) || has_colon(Base.tail(idx))
-
-# TODO: better error (or just handle correctly) when user tries to index with a range like a:b
-# The only kind of slicing we support is dropping a dimension with colons
-function Base.getindex(A::PowerNetworkMatrix, row, column)
-    #=
-    This is old code when we accepted idx... instead of row, column
-    if has_colon(idx)
-        PTDF(A.data[to_index(A,idx...)...], (ax for (i,ax) in enumerate(A.axes) if idx[i] == Colon())...)
+    BA = SparseArrays.sparse(BA_I, BA_J, BA_V)
+
+    return BA
+end
+
+# ABA matrix evaluation ######################################################
+function calculate_ABA_matrix(
+    A::SparseArrays.SparseMatrixCSC{Int8, Int32},
+    BA::SparseArrays.SparseMatrixCSC{T, Int32} where {T <: Union{Float32, Float64}},
+    slack_positions::Vector{Int64})
+    return A[:, setdiff(1:end, slack_positions[1])]' * BA
+end
+
+# PTDF evaluation ############################################################
+function calculate_PTDF_matrix_KLU(branches, nodes::Vector{PSY.Bus},
+    dist_slack::Vector{Float64})
+    buscount = length(nodes)
+    linecount = length(branches)
+
+    A, slack_positions = calculate_A_matrix(branches, nodes)
+    bus_lookup = _make_ax_ref(nodes)
+    BA = calculate_BA_matrix(branches, slack_positions, bus_lookup)
+    ABA = calculate_ABA_matrix(A, BA, slack_positions)
+    K = klu(ABA)
+    Ix = Matrix(1.0LinearAlgebra.I, size(ABA, 1), size(ABA, 1))
+    ABA_inv = zeros(Float64, size(ABA, 1), size(ABA, 2))
+    ldiv!(ABA_inv, K, Ix)
+
+    slack_position = slack_positions[1]
+    if dist_slack[1] == 0.1 && length(dist_slack) == 1
+        S = BA * ABA_inv
+        @views S =
+            hcat(S[:, 1:(slack_position - 1)], zeros(linecount), S[:, slack_position:end])
+    elseif dist_slack[1] != 0.1 && length(dist_slack) == buscount
+        @info "Distributed bus"
+        S = BA * ABA_inv
+        @views S =
+            hcat(S[:, 1:(slack_position - 1)], zeros(linecount), S[:, slack_position:end])
+        slack_array = dist_slack / sum(dist_slack)
+        slack_array = reshape(slack_array, buscount, 1)
+        S = S - (S * slack_array) * ones(1, buscount)
+    elseif length(slack_position) == 0
+        @warn("Slack bus not identified in the Bus/Nodes list, can't build PTDF")
+        S = Array{Float64, 2}(undef, linecount, buscount)
     else
-        return A.data[to_index(A,idx...)...]
+        @assert false
     end
-    =#
-    return A.data[to_index(A, row, column)...]
-end
-Base.getindex(A::PowerNetworkMatrix, idx::CartesianIndex) = A.data[idx]
-Base.setindex!(A::PowerNetworkMatrix, v, idx...) = A.data[to_index(A, idx...)...] = v
-Base.setindex!(A::PowerNetworkMatrix, v, idx::CartesianIndex) = A.data[idx] = v
 
-Base.IndexStyle(::Type{PowerNetworkMatrix}) = IndexAnyCartesian()
+    return S, A
```

Comment:
no need to return A here

Reply:
we agreed to keep them both for the momen